File: packages/vela-sdk/src/vela_sdk/fastmcp/elicitation.py
"""Elicitation Service — maps CaptureDefinition to FastMCP ctx.elicit() calls."""

from typing import Any

from fastmcp.server.elicitation import (
    AcceptedElicitation,
    CancelledElicitation,
    DeclinedElicitation,
)
from pydantic import BaseModel, Field, create_model

from vela_sdk.schemas.workflow import CaptureDefinition


def _titled_scalar(title: str, python_type: type) -> type:
    """Create a Pydantic model with a titled 'value' field.

    This ensures the MCP elicitation schema includes a human-readable title
    for the input field instead of the generic "value" label.
    """
    return create_model(
        "ElicitResponse",
        value=(python_type, Field(title=title)),
    )


class ElicitationService:
    """Stateless service for building and processing elicitation requests."""

    @staticmethod
    def needs_elicitation(
        captures: list[CaptureDefinition], state_data: dict
    ) -> list[CaptureDefinition]:
        """Filter captures that need elicitation.

        - elicit="always" -> always
        - elicit="if_missing" -> only if key not in state_data
        - elicit="never" -> never
        """
        result: list[CaptureDefinition] = []
        for cap in captures:
            if cap.elicit == "always":
                result.append(cap)
            elif cap.elicit == "if_missing" and cap.key not in state_data:
                result.append(cap)
            # elicit="never" -> skip
        return result

    @staticmethod
    def build_response_type(
        capture: CaptureDefinition,
    ) -> type | list[str] | dict[str, dict[str, str]] | list[list[str]] | None:
        """Map capture.input to FastMCP elicit response_type."""
        input_type = capture.input
        title = capture.label or capture.key

        if input_type == "confirm":
            return _titled_scalar(title, bool)
        elif input_type == "text":
            return _titled_scalar(title, str)
        elif input_type == "number":
            return _titled_scalar(title, int)
        elif input_type == "boolean":
            return _titled_scalar(title, bool)
        elif input_type == "select":
            if capture.options:
                has_labels = any(o.label != o.key for o in capture.options)
                if has_labels:
                    return {
                        o.key: {"title": o.label} for o in capture.options
                    }
                return [o.key for o in capture.options]
            return _titled_scalar(title, str)
        elif input_type == "multi-select":
            if capture.options:
                has_labels = any(o.label != o.key for o in capture.options)
                if has_labels:
                    return [
                        {o.key: {"title": o.label} for o in capture.options}
                    ]
                return [[o.key for o in capture.options]]
            return _titled_scalar(title, str)
        elif input_type is None:
            return _titled_scalar(title, str)
        else:
            return _titled_scalar(title, str)

    @staticmethod
    def build_message(capture: CaptureDefinition) -> str:
        """Build the elicit message from capture definition."""
        label = capture.label or capture.key

        parts = [label]

        if capture.placeholder:
            parts.append(f"(e.g. {capture.placeholder})")

        if capture.default is not None:
            parts.append(f"[default: {capture.default}]")

        return " ".join(parts)

    @staticmethod
    def process_result(
        capture: CaptureDefinition, result: Any
    ) -> tuple[str, Any] | None:
        """Process an ElicitationResult.

        - AcceptedElicitation -> (capture.key, result.data)
        - DeclinedElicitation/CancelledElicitation -> None
        """
        if isinstance(result, AcceptedElicitation):
            data = result.data
            # Extract value from Pydantic model (used for titled scalar fields)
            if isinstance(data, BaseModel) and hasattr(data, "value"):
                data = data.value
            return (capture.key, data)
        if isinstance(result, (DeclinedElicitation, CancelledElicitation)):
            return None
        return None

File: packages/vela-sdk-ts/src/mcp/elicitation.ts
/**
 * Elicitation Service — maps CaptureDefinition to MCP elicit JSON Schema calls.
 *
 * Uses raw JSON Schema (not Zod) since MCP protocol expects JSON Schema objects.
 */

import type { CaptureDefinition } from "../schemas/workflow.js";
import type { ElicitResult } from "./mcp-server.js";

// ---------------------------------------------------------------------------
// ElicitationService
// ---------------------------------------------------------------------------

export class ElicitationService {
  /**
   * Filter captures that still need elicitation.
   *
   * - `elicit="always"` -> always
   * - `elicit="if_missing"` -> only if key not in stateData
   * - `elicit="never"` -> skip
   */
  static needsElicitation(
    captures: CaptureDefinition[],
    stateData: Record<string, unknown>,
  ): CaptureDefinition[] {
    const result: CaptureDefinition[] = [];
    for (const cap of captures) {
      if (cap.elicit === "always") {
        result.push(cap);
      } else if (cap.elicit === "if_missing" && !(cap.key in stateData)) {
        result.push(cap);
      }
      // elicit="never" -> skip
    }
    return result;
  }

  /**
   * Build a JSON Schema object for the MCP elicit call based on the capture's
   * input type.
   *
   * Mapping:
   * - text / null / unknown -> `{ type: "string" }`
   * - number -> `{ type: "number" }`
   * - boolean / confirm -> `{ type: "boolean" }`
   * - select -> `{ type: "string", enum: [...] }`
   * - multi-select -> `{ type: "array", items: { type: "string", enum: [...] } }`
   */
  static buildRequestedSchema(
    capture: CaptureDefinition,
  ): Record<string, unknown> {
    const title = capture.label ?? capture.key;
    const inputType = capture.input;

    if (inputType === "confirm" || inputType === "boolean") {
      return {
        type: "object",
        properties: {
          value: { type: "boolean", title },
        },
        required: ["value"],
      };
    }

    if (inputType === "number") {
      return {
        type: "object",
        properties: {
          value: { type: "number", title },
        },
        required: ["value"],
      };
    }

    if (inputType === "select") {
      if (capture.options && capture.options.length > 0) {
        const enumValues = capture.options.map((o) => o.key);
        const hasLabels = capture.options.some((o) => o.label !== o.key);
        const prop: Record<string, unknown> = {
          type: "string",
          title,
          enum: enumValues,
        };
        if (hasLabels) {
          const descriptions: Record<string, string> = {};
          for (const o of capture.options) {
            descriptions[o.key] = o.label;
          }
          prop["x-enum-descriptions"] = descriptions;
        }
        return {
          type: "object",
          properties: { value: prop },
          required: ["value"],
        };
      }
      return {
        type: "object",
        properties: { value: { type: "string", title } },
        required: ["value"],
      };
    }

    if (inputType === "multi-select") {
      if (capture.options && capture.options.length > 0) {
        const enumValues = capture.options.map((o) => o.key);
        const hasLabels = capture.options.some((o) => o.label !== o.key);
        const itemSchema: Record<string, unknown> = {
          type: "string",
          enum: enumValues,
        };
        if (hasLabels) {
          const descriptions: Record<string, string> = {};
          for (const o of capture.options) {
            descriptions[o.key] = o.label;
          }
          itemSchema["x-enum-descriptions"] = descriptions;
        }
        return {
          type: "object",
          properties: {
            value: { type: "array", title, items: itemSchema },
          },
          required: ["value"],
        };
      }
      return {
        type: "object",
        properties: { value: { type: "string", title } },
        required: ["value"],
      };
    }

    // text, null, or unknown -> string
    return {
      type: "object",
      properties: {
        value: { type: "string", title },
      },
      required: ["value"],
    };
  }

  /**
   * Build a human-readable elicit message from a capture definition.
   */
  static buildMessage(capture: CaptureDefinition): string {
    const label = capture.label ?? capture.key;
    const parts: string[] = [label];

    if (capture.placeholder) {
      parts.push(`(e.g. ${capture.placeholder})`);
    }

    if (capture.default !== undefined && capture.default !== null) {
      parts.push(`[default: ${capture.default}]`);
    }

    return parts.join(" ");
  }

  /**
   * Process an ElicitResult.
   *
   * - `action === "accept"` -> `[capture.key, extractedValue]`
   * - `action === "decline" | "cancel"` -> `null`
   */
  static processResult(
    capture: CaptureDefinition,
    result: ElicitResult,
  ): [string, unknown] | null {
    if (result.action === "accept" && result.content) {
      // Extract value from the wrapper object
      const data = result.content;
      if ("value" in data) {
        return [capture.key, data["value"]];
      }
      // Fallback: return entire content
      return [capture.key, data];
    }
    return null;
  }
}
